core/trino-main/src/test/java/io/trino/sql/planner/TestArraySortAfterArrayDistinctRewriter.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import io.trino.sql.ExpressionTestUtils;
+import io.trino.sql.planner.assertions.PlanMatchPattern;
+import io.trino.sql.planner.iterative.rule.ArraySortAfterArrayDistinct;
+import io.trino.sql.planner.iterative.rule.test.BaseRuleTest;
+import io.trino.sql.planner.iterative.rule.test.PlanBuilder;
+import io.trino.sql.planner.plan.Assignments;
+import io.trino.sql.tree.Expression;
+import org.testng.annotations.Test;
+
+import java.util.Map;
+
+import static io.trino.sql.planner.assertions.PlanMatchPattern.project;
+import static io.trino.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestArraySortAfterArrayDistinctRewriter
+        extends BaseRuleTest
+{
+    @Test
+    public void basic()
+    {
+        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY['a']))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY['a']))");
+    }
+
+//    @Test
+//    // TODO: should figure out why this is failing, although the plan seems correct
+//    public void withLambda()
+//    {
+//        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY[1], (x, y) -> IF(x < y, -1, IF(x > y, 1, 0 ))))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY[1]), (x, y) -> IF(x < y, -1, IF(x > y, 1, 0 )))");
+//    }
+
+    @Test
+    public void withRecursion()
+    {
+        test("ARRAY_DISTINCT(ARRAY_SORT(ARRAY_DISTINCT(ARRAY_SORT(ARRAY['a']))))", "ARRAY_SORT(ARRAY_DISTINCT(ARRAY_SORT(ARRAY_DISTINCT(ARRAY['a']))))");
```

Comment:
Here we have the outcome of single application of the rule.
If you added this test case to `TestLogicalPlanner`, we could see the function call chain fully sorted.

Reply:
I need some pointers here.

I've tried to add some simple test here:

```
    @Test
    public void testArraySortAfterArrayDistinctRecursive()
    {
        assertPlan("SELECT ARRAY_SORT(items) from (VALUES (ARRAY ['cat', 'dog', 'dog'])) t(items)",
                output(project(ImmutableMap.of("output", expression("ARRAY_SORT(items)")),
                        values())));
    }
```

However this results in following plan:

```
Output[_col0]
│   Layout: [array_sort:array(varchar(3))]
│   _col0 := array_sort
└─ Values
       Layout: [array_sort:array(varchar(3))]
       ("$literal$"(from_base64('DgAAAFZBUklBQkxFX1dJRFRIAwAAAAMAAAAGAAAACQAAAAAJAAAAY2F0ZG9nZG9n')))

```

How can I express such a plan?

Reply:
In your example, first a VALUES-related optimization `MergeProjectWithValues` fires, and the `array_sort` call is inlined into VALUES, and then the expression in VALUES is evaluated into literal. To avoid this, try using a non-deterministic expression (it shouldn't be folded into a literal):
```
VALUES array_distinct(array_sort(array_distinct(array_sort(ARRAY[0e0, 1e0, rand()]))))
```

Reply:
Great, I will try it out!

Reply:
Getting a bit further. However failing in ValuesMatcher as it seems the ARRAY_CONSTRUCTOR doesn't match up (failing at `!expectedRows.equals(valuesNode.getRows())`)

```
    @Test
    public void testArraySortAfterArrayDistinctRecursive2()
    {
        FunctionCall randomFunction = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("rand"), ImmutableList.of()).toQualifiedName(),
                ImmutableList.of());

        FunctionCall arrayConstructor = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("array_constructor"), fromTypes(DOUBLE)).toQualifiedName(),
                ImmutableList.of(new DoubleLiteral("0e0"), new DoubleLiteral("1e0"), randomFunction));

        assertPlan("SELECT * FROM (VALUES (ARRAY[0e0, 1e0, rand()])) T(x)",
                output(values(
                                ImmutableList.of("x"),
                                ImmutableList.of(ImmutableList.of(arrayConstructor)))));
    }
```

Reply:
That's because the resolved function for `array_constructor` doesn't match. If you replace it with the below  (it has `fromTypes(DOUBLE, DOUBLE, DOUBLE)`), it passes.

```
FunctionCall arrayConstructor = new FunctionCall(
                getQueryRunner().getMetadata().resolveFunction(QualifiedName.of("array_constructor"), fromTypes(DOUBLE, DOUBLE, DOUBLE)).toQualifiedName(),
                ImmutableList.of(new DoubleLiteral("0e0"), new DoubleLiteral("1e0"), randomFunction));
```